src/cryptographic_primitives/twoparty/dh_key_exchange.rs

```diff
@@ -0,0 +1,204 @@
+    ) -> Result<Party1SecondMessage, ProofError> {
+        DLogProof::verify(&proof)?;
+        Ok(Party1SecondMessage {
+            pk_commitment_blind_factor: first_message.pk_commitment_blind_factor.clone(),
+            zk_pok_blind_factor: first_message.zk_pok_blind_factor.clone(),
+            public_share: first_message.public_share.clone(),
+            d_log_proof: first_message.d_log_proof.clone(),
+        })
+    }
+}
+
+impl Party2FirstMessage {
+    pub fn create() -> Party2FirstMessage {
+        let base: GE = ECPoint::new();
+        let sk: FE = ECScalar::new_random();
+        let pk = base.scalar_mul(&sk.get_element());
+        Party2FirstMessage {
+            d_log_proof: DLogProof::prove(&sk),
+            public_share: pk,
+            secret_share: sk,
+        }
+    }
+}
+
+impl Party2SecondMessage {
+    pub fn verify_commitments_and_dlog_proof(
+        party_one_pk_commitment: &BigInt,
+        party_one_zk_pok_commitment: &BigInt,
+        party_one_zk_pok_blind_factor: &BigInt,
+        party_one_public_share: &GE,
+        party_one_pk_commitment_blind_factor: &BigInt,
+        party_one_d_log_proof: &DLogProof,
+    ) -> Result<Party2SecondMessage, ProofError> {
+        let mut flag = true;
+        match party_one_pk_commitment
+            == &HashCommitment::create_commitment_with_user_defined_randomness(
+                &party_one_public_share.get_x_coor_as_big_int(),
+                &party_one_pk_commitment_blind_factor,
+            ) {
+            false => flag = false,
+            true => flag = flag,
+        };
+        match party_one_zk_pok_commitment
+            == &HashCommitment::create_commitment_with_user_defined_randomness(
+                &party_one_d_log_proof
+                    .pk_t_rand_commitment
+                    .get_x_coor_as_big_int(),
+                &party_one_zk_pok_blind_factor,
+            ) {
+            false => flag = false,
+            true => flag = flag,
+        };
+        assert!(flag);
+        DLogProof::verify(&party_one_d_log_proof)?;
+        Ok(Party2SecondMessage {})
+    }
+}
+
+pub fn compute_pubkey_party1(
+    party_one_first_message: &Party1FirstMessage,
+    party_two_first_message: &Party2FirstMessage,
+) -> GE {
+    let pubkey = party_two_first_message.public_share.clone();
+    pubkey.scalar_mul(&party_one_first_message.secret_share.get_element())
+}
+pub fn compute_pubkey_party2(
+    party_two_first_message: &Party2FirstMessage,
+    party_one_first_message: &Party1FirstMessage,
+) -> GE {
+    let pubkey = party_one_first_message.public_share.clone();
+    pubkey.scalar_mul(&party_two_first_message.secret_share.get_element())
+}
+pub fn compute_pubkey(secret_share: &FE, public_share: &GE) -> GE {
+    let pubkey = public_share.clone();
+    pubkey.scalar_mul(&secret_share.get_element())
+}
+#[cfg(test)]
+mod tests {
+    use cryptographic_primitives::twoparty::dh_key_exchange::*;
+
+    #[test]
+    fn test_full_key_gen() {
```

Comment:
I'm wondering if the interface could be changed to avoid this picking of fields, which might cause some confusion or worse down the line. What if eg `verify_commitments_and_dlog_proof` simply took `party_one_first_message` and `party_one_second_message` as arguments?